futurewei-cloud/arktos
# kubelet checks network readiness

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Kubelet syncPod workflow has one more step to check pod's network-readiness annotation value. 

As depicted in [multi-tenancy networking design proposal](https://github.com/futurewei-cloud/arktos/blob/master/docs/design-proposals/multi-tenancy/multi-tenancy-network.md), pod has arktos.futurewei.com/network-readiness annotation served as signal that external network component indicates the preparation of networking for the pod has been done, and pod is able to start. It is necessary for kubelet to add one more check in the syncPod workflow to ensure the signal is not blocking.

This is one feature needed in kubelet to support multi-tenant networking.  

**How to verify on local dev env**:
1. start Arktos system by ```hack/arktos-up.sh```
2. create a pod that has network-readiness=false, by ```kubectl create -f pod-net-not-ready.yaml```, with the yaml file 
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx-system-n-n-r
  annotations:
    "arktos.futurewei.com/network-readiness": "false"
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
    - containerPort: 80
```
3. observe the pod just created won't be able to start;
```
$ cluster/kubectl.sh get pods nginx-system-n-n-r
nginx-system-n-n-r   3277677723223307134   0/1     ContainerCreating   0          10m
```
4. change network-readiness value to "true" using ```kubectl edit pod nginx-system-n-n-r```;
5. shortly this pod turns into Running state
```
$ cluster/kubectl.sh get pods nginx-system-n-n-r
nginx-system-n-n-r          3277677723223307134   1/1     Running   0          12m
```

**Does this PR introduce a user-facing change?**:
Yes
```release-note
pods annotated by key of "arktos.futurewei.com/network-readiness" and value if not "true" shall be put on hold to start at the work node, until the condition is cleared.
```
